mmdet/models/detectors/semi_base.py

```diff
@@ -0,0 +1,297 @@
+                The dict of multi-branch data samples.
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        losses = dict()
+        losses.update(**self.gt_loss(multi_batch_inputs['sup'],
+                                     multi_batch_data_samples['sup']))
+        origin_pseudo_instances, batch_info = self.get_pseudo_instances(
+            multi_batch_inputs['unsup_teacher'],
+            multi_batch_data_samples['unsup_teacher'])
+        multi_batch_data_samples[
+            'unsup_student'] = self.project_pseudo_instances(
+                origin_pseudo_instances,
+                multi_batch_data_samples['unsup_student'])
+        losses.update(**self.pseudo_loss(
+            multi_batch_inputs['unsup_student'],
+            multi_batch_data_samples['unsup_student'], batch_info))
+        return losses
+
+    def gt_loss(self, batch_inputs: Tensor,
+                batch_data_samples: SampleList) -> dict:
+        """Calculate losses from a batch of inputs and ground-truth data
+        samples.
+
+        Args:
+            batch_inputs (Tensor): Input images of shape (N, C, H, W).
+                These should usually be mean centered and std scaled.
+            batch_data_samples (List[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`.
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        gt_loss = {
+            'sup_' + k: v
+            for k, v in self.weight(
+                self.student.loss(batch_inputs, batch_data_samples),
+                self.semi_train_cfg.get('sup_weight', 1.)).items()
+        }
+        return gt_loss
+
+    def pseudo_loss(self,
+                    batch_inputs: Tensor,
+                    batch_data_samples: SampleList,
+                    batch_info: Optional[dict] = None) -> dict:
+        """Calculate losses from a batch of inputs and pseudo data samples.
+
+        Args:
+            batch_inputs (Tensor): Input images of shape (N, C, H, W).
+                These should usually be mean centered and std scaled.
+            batch_data_samples (List[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`,
+                which are `pseudo_instance` or `pseudo_panoptic_seg`
+                or `pseudo_sem_seg` in fact.
+            batch_info (dict): Batch information of teacher model
+                forward propagation process. Defaults to None.
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        for data_samples in batch_data_samples:
+            if data_samples.gt_instances.bboxes.shape[0] > 0:
+                data_samples.gt_instances = data_samples.gt_instances[
```

Comment:
Why not put it in `filter_pseudo_instances`？

Reply:
Subclasses need filtering pseudo instances by different scores.

Reply:
But all semi-model need filter too small or invalid pseudo instances.